tmtest/tmtest.go

```diff
@@ -49,3 +58,110 @@ func RunTendermint(ctx context.Context, t *testing.T, home string) (cleanup func
 		cmd.Wait()
 	}
 }
+
+// RunBnsd is like RunTendermint, just executes the bnsd executable, assuming a prepared home directory
+func RunBnsd(ctx context.Context, t *testing.T, home string) (cleanup func()) {
+	t.Helper()
+
+	bnsdpath, err := exec.LookPath("bnsd")
+	if err != nil {
+		if os.Getenv("FORCE_TM_TEST") != "1" {
+			t.Skip("Bnsd binary not found. Set FORCE_TM_TEST=1 to fail this test.")
+		} else {
+			t.Fatalf("Bnsd binary not found. Do not set FORCE_TM_TEST=1 to skip this test.")
+		}
+	}
+
+	cmd := exec.CommandContext(ctx, bnsdpath, "-home", home, "start")
+	// log tendermint output for verbose debugging....
+	if os.Getenv("TM_DEBUG") != "" {
+		cmd.Stdout = os.Stderr
+		cmd.Stderr = os.Stderr
+	}
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("Bnsd process failed: %s", err)
+	}
+
+	// Give tendermint time to setup.
+	time.Sleep(2 * time.Second)
+	t.Logf("Running %s pid=%d", bnsdpath, cmd.Process.Pid)
+
+	// Return a cleanup function, that will wait for the tendermint to stop.
+	//nolint
+	return func() {
```

Comment:
`RunBnsd` takes `context.Context` instance. Instead of returning a cleanup function, I think it might be better to use given context.

```go
// No return func() { kill & stop }
// This function returns nothing and does cleanup when context is done.
go func () {
    <-ctx.Done()
    cmd.Process.Kill()
}()
```
The only difference is that there is no `cmd.Wait` as this is asynchronous stop. Do we need to keep it synchronous or many Tendermint instances are able to run at the same time?

Reply:
One thing that would be beneficial for are tests.

We already have some flakiness which shows itself time to time, where previous tendermint
is still running when spinning up a new one. 

If one needs an async api for that - just do `go func() {RunBnsd(...); whatever error handling and cleanup}()

Reply:
Although I do like the async cleanup via context, I think having explicit synchronous control over this is important for testing and that ultimately is a utility for writing tests.

Reply:
I liked your point on context, @husio , but yes, I agree with @ruseinov that it is important to ensure the tendermint process is killed *before* the next test starts. With goroutines, we add a race condition.

I think there is no way to do this with context as cancelling a context is by nature not blocked by processing the actions that are cancelled.